### PR TITLE
Issue 4165 - Don't apply rootDN access control plugin to UNIX connections

### PR DESCRIPTION
**Pull request by yrro**
Closes: #4165

### CODE REVIEW

**Comment by Firstyear** [comment]
I think this all looks good to me - did you want to take this out of being a draft @yrro ? 

@mreynolds389 I think you may want to have a review on this too, I seem to recall that you have looked at this plugin in the past ....

**Comment by yrro** [comment]
Thanks for taking a look. I wanted to test this first, in doing so I found that I'd totally broken the plugin - I was only testing for PR_AF_INET, but of course I should be testing for IPv6 as well. So I decided to flip the test around, and return early iff the connection is PF_AF_LOCAL.

Anyway, I think the code works now. Test setup:

```
$ sudo dsconf wintermute plugin root-dn show
dn: cn=RootDN Access Control,cn=plugins,cn=config
cn: RootDN Access Control
nsslapd-plugin-depends-on-type: database
nsslapd-pluginDescription: RootDN Access Control plugin
nsslapd-pluginEnabled: on
nsslapd-pluginId: RootDN Access Control
nsslapd-pluginInitfunc: rootdn_init
nsslapd-pluginPath: librootdn-access-plugin
nsslapd-pluginType: internalpreoperation
nsslapd-pluginVendor: 389 Project
nsslapd-pluginVersion: 1.4.4.11
objectClass: top
objectClass: nsSlapdPlugin
objectClass: rootDNPluginConfig
rootdn-allow-ip: 127.0.0.1
```

Before applying the patch:

```
$ dpkg-query --show 389-ds-base
389-ds-base	1.4.4.17-1+b1

$ ldapwhoami -H ldap://127.0.0.1 -D 'cn=Directory Manager' -w dmpwdmpw
dn: cn=directory manager

$ ldapwhoami -H ldap://10.0.2.15 -D 'cn=Directory Manager' -w dmpwdmpw
ldap_bind: Server is unwilling to perform (53)
	additional info: RootDN access control violation

$ ldapwhoami -H ldapi://%2frun%2fslapd-wintermute.socket -D 'cn=Directory Manager' -w dmpwdmpw
ldap_bind: Server is unwilling to perform (53)
	additional info: RootDN access control violation
```

After applying the patch:

```
$ dpkg-query --show 389-ds-base
389-ds-base	1.4.4.17-1sam1

$ ldapwhoami -H ldap://127.0.0.1 -D 'cn=Directory Manager' -w dmpwdmpw
dn: cn=directory manager

$ ldapwhoami -H ldap://10.0.2.15 -D 'cn=Directory Manager' -w dmpwdmpw
ldap_bind: Server is unwilling to perform (53)
	additional info: RootDN access control violation

$ ldapwhoami -H ldapi://%2frun%2fslapd-wintermute.socket -D 'cn=Directory Manager' -w dmpwdmpw
dn: cn=directory manager
```

File: ldap/servers/slapd/pblock.c [hunk 1]
```diff
@@ -492,6 +492,20 @@ slapi_pblock_get(Slapi_PBlock *pblock, int arg, void *value)
         (*(PRNetAddr **) value) = pblock->pb_conn->cin_addr_aclip;
         pthread_mutex_unlock(&(pblock->pb_conn->c_mutex));
         break;
+    case SLAPI_CONN_SERVERNETTYPE:
```
Comment (yrro) [inline ldap/servers/slapd/pblock.c]: "type" is a bit of a vague term. Perhaps `SLAPI_CONN_SERVERNETADDRESSFAMILY` would be better?

File: ldap/servers/plugins/rootdn_access/rootdn_access.c [hunk 2]
```diff
@@ -513,6 +514,15 @@ rootdn_check_access(Slapi_PBlock *pb)
             return -1;
         }
     }
+
+    /*
+     *  Remaining checks are only relevant for AF_INET/AF_INET6 connections
+     */
+    slapi_pblock_get(pb, SLAPI_CONN_SERVERNETTYPE, &server_net_type);
```
Comment (yrro) [inline ldap/servers/plugins/rootdn_access/rootdn_access.c]: Actually we don't need to add a whole new request type. The data we need is already available if we ask for `SLAPI_CONN_SERVERNETADDR`, we have only to check the address family with `PR_NetAddrFamily` here.

**Review by Firstyear — APPROVED** [review]
Even better :)

**Comment by mreynolds389** [comment]
> I think this all looks good to me - did you want to take this out of being a draft @yrro ?
> 
> @mreynolds389 I think you may want to have a review on this too, I seem to recall that you have looked at this plugin in the past ....

I wrote the plugin :-)  This actually solves an issue I was worried about!  Well I want to add some form of "account lock" to the Root DN, but there should still be a way to perform updates as Root DN, so always allowing LDAPI makes sense and addresses that concern.  

So ACK from me!  Thanks @yrro for your contribution.  Before merging lets see if this breaks any CI tests first.  Starting tests...

**Comment by yrro** [comment]
Just squashed down to a single commit.

**Comment by mreynolds389** [comment]
The tests have passed, please update the commit message to set the reviewers name(s) and then we can merge this!

**Comment by yrro** [comment]
Done. Shall I do a separate PR against the master branch?

**Comment by mreynolds389** [comment]
> Done. Shall I do a separate PR against the master branch?

No, I will cherry pick it to the other branches after the merge.  Thanks again for the contribution @yrro !

Merging...

**Comment by yrro** [comment]
Awesome. Thanks both!

**Pull request merged**